chronos/streams/chunkstream.nim

```diff
@@ -68,161 +87,116 @@ proc setChunkSize(buffer: var openarray[byte], length: int64): int =
     buffer[c] = byte(0x0D)
     buffer[c + 1] = byte(0x0A)
-    result = c + 2
+    c + 2
 
 proc chunkedReadLoop(stream: AsyncStreamReader) {.async.} =
   var rstream = cast[ChunkedStreamReader](stream)
   var buffer = newSeq[byte](1024)
   rstream.state = AsyncStreamState.Running
 
-  try:
-    while true:
+  while true:
+    try:
       # Reading chunk size
-      var ruFut1 = awaitne rstream.rsource.readUntil(addr buffer[0], 1024, CRLF)
-      if ruFut1.failed():
-        rstream.error = ruFut1.error
-        rstream.state = AsyncStreamState.Error
-        break
-
-      let length = ruFut1.read()
-      var chunksize = getChunkSize(buffer.toOpenArray(0,
-                                                      length - len(CRLF) - 1))
+      let res = await rstream.rsource.readUntil(addr buffer[0], 1024, CRLF)
+      var chunksize = getChunkSize(buffer.toOpenArray(0, res - len(CRLF) - 1))
 
       if chunksize == 0xFFFF_FFFF_FFFF_FFFF'u64:
-        rstream.error = newProtocolError()
+        rstream.error = newChunkedProtocolError()
         rstream.state = AsyncStreamState.Error
-        break
       elif chunksize > 0'u64:
         while chunksize > 0'u64:
           let toRead = min(int(chunksize), rstream.buffer.bufferLen())
-          var reFut2 = awaitne rstream.rsource.readExactly(
-                                             rstream.buffer.getBuffer(), toRead)
-          if reFut2.failed():
-            rstream.error = reFut2.error
-            rstream.state = AsyncStreamState.Error
-            break
-
+          await rstream.rsource.readExactly(rstream.buffer.getBuffer(), toRead)
           rstream.buffer.update(toRead)
           await rstream.buffer.transfer()
           chunksize = chunksize - uint64(toRead)
 
-        if rstream.state != AsyncStreamState.Running:
-          break
+        if rstream.state == AsyncStreamState.Running:
+          # Reading chunk trailing CRLF
+          await rstream.rsource.readExactly(addr buffer[0], 2)
 
-        # Reading chunk trailing CRLF
-        var reFut3 = awaitne rstream.rsource.readExactly(addr buffer[0], 2)
-        if reFut3.failed():
-          rstream.error = reFut3.error
-          rstream.state = AsyncStreamState.Error
-          break
-
-        if buffer[0] != CRLF[0] or buffer[1] != CRLF[1]:
-          rstream.error = newProtocolError()
-          rstream.state = AsyncStreamState.Error
-          break
+          if buffer[0] != CRLF[0] or buffer[1] != CRLF[1]:
+            rstream.error = newChunkedProtocolError()
+            rstream.state = AsyncStreamState.Error
       else:
         # Reading trailing line for last chunk
-        var ruFut4 = awaitne rstream.rsource.readUntil(addr buffer[0],
-                                                       len(buffer), CRLF)
-        if ruFut4.failed():
-          rstream.error = ruFut4.error
-          rstream.state = AsyncStreamState.Error
-          break
-
+        discard await rstream.rsource.readUntil(addr buffer[0],
```

Comment:
Shouldn't this follow the code above trying to read exactly two bytes with `rstream.rsource.readExactly(addr buffer[0], 2)` and then verifying that we got `CRLF`? 

Also, why aren't we breaking the loop after the last chunk?

Reply:
@zah according to specification https://tools.ietf.org/html/rfc2616#section-3.6.1
```
last-chunk     = 1*("0") [ chunk-extension ] CRLF
chunk-extension= *( ";" chunk-ext-name [ "=" chunk-ext-val ] )
chunk-ext-name = token
chunk-ext-val  = token | quoted-string
```
So as you can see right after `0` there could be a lot of text followed by CRLF.

@zah we are breaking loop right after we got last chunk, it is made by setting 
https://github.com/status-im/nim-chronos/pull/153/files#diff-2a604231f48191766b227ea45d864d890caa5cf396b627bdb751b0061f47b9f4R125
stream state to `AsyncStreamState.Finished` and after exception handlers we are exiting loop.

Reply:
I was going to raise the question of handling chunk extensions today. Aren't they allowed on every chunk, not just the last one? Right now, we expect to read a chunk size until CRLF is reached, but if chunk extensions are used, the chunk size buffer will end up containing the chunk extensions string and `getChunkSize` will detect this as an error. So my assumption was that you have postponed adding the support for chunk extensions.

Reply:
Regarding breaking, I've noticed that you are setting the `Finished` state, but the loop terminating condition on line 136:
https://github.com/status-im/nim-chronos/pull/153/files#diff-2a604231f48191766b227ea45d864d890caa5cf396b627bdb751b0061f47b9f4R136

... is checking only for the `Stopped` and `Error` states. What am I missing?

Reply:
Fixed.